plugins/flashrom-i2c/fu-plugin-flashrom-i2c.c

```diff
@@ -0,0 +1,198 @@
+ * found in the LICENSE file.
+ */
+
+#include <fu-plugin.h>
+#include <fu-plugin-vfuncs.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "config.h"
+#include "fu-hash.h"
+#include "fu-i2c-device.h"
+
+#define DEVICE_NAME_PREFIX "i2c-"
+#define HID_LENGTH 8
+#define I2C_PATH_REGEX "^i2c-[0-9]+$"
+#define DEVICE_GUID_SOURCE_FORMAT "FLASHROM-I2C\\VEN_%s&amp;DEV_%s"
+#define SYMLINK_LEN_MAX 128
+
+static gboolean i2c_match_regex (const char *target, GError **error)
+{
+	GRegex *regex = NULL;
+	gboolean result;
+
+	regex = g_regex_new (I2C_PATH_REGEX, 0, 0, error);
+	if (regex == NULL)
+		return FALSE;
+
+	result = g_regex_match_all_full (regex, target, -1, 0, 0, NULL, error);
+	g_regex_unref(regex);
+	return result;
+}
+
+static gint get_i2c_bus_number_from_path (gchar *device_path, GError **error)
+{
+	gint bus_no = -1;
+	g_autofree gchar *device_symlink = NULL;
+	g_auto(GStrv) names = NULL;
+
+	device_symlink = g_file_read_link (device_path, error);
+	if (device_symlink == NULL)
+		return bus_no;
+
+	names = g_strsplit (device_symlink, "/", -1);
+	for (guint i = 0; names[i] != NULL; i++) {
+		if (i2c_match_regex (names[i], error)) {
+			bus_no = g_ascii_strtoll (
+				&names[i][strlen (DEVICE_NAME_PREFIX)],
+				NULL,
+				10);
+			break;
+		}
+	}
+
+	return bus_no;
+}
+
+static gchar *get_i2c_device_guid (const gchar *pid, const gchar *vid)
+{
+	gchar *guid_source = NULL;
+	guid_source = g_strdup_printf (DEVICE_GUID_SOURCE_FORMAT, pid, vid);
+	return fwupd_guid_hash_string (guid_source);
+}
+
+static gboolean add_i2c_device (FuPlugin *plugin, const gchar *i2c_device_dir,
+				const gchar *i2c_name, GError **error)
+{
+	gint bus_no;
+	const gchar *quirk_programmer_name;
+	const gchar *quirk_device_name;
+	const gchar *quirk_device_protocol;
+	const gchar *quirk_vendor_name;
+	g_autofree gchar *hw_id = NULL;
+	g_autofree gchar *quirk_key = NULL;
+	g_autofree gchar *device_path = NULL;
+	g_autofree gchar *device_vid = NULL;
+	g_autofree gchar *device_pid = NULL;
+	g_autofree gchar *device_guid = NULL;
+	g_autofree gchar *vendor_id = NULL;
+	g_autofree gchar *physical_id = NULL;
+	g_autoptr (FuI2cDevice) dev = NULL;
+
+	hw_id = g_strndup (&i2c_name[strlen (DEVICE_NAME_PREFIX)],
+			     HID_LENGTH);
+	quirk_key = g_strdup_printf ("HwId=%s", hw_id);
+
+	quirk_programmer_name = fu_plugin_lookup_quirk_by_id (
+		plugin, quirk_key, PROGRAMMER_NAME);
+	quirk_device_name = fu_plugin_lookup_quirk_by_id (
+		plugin, quirk_key, DEVICE_NAME);
+	quirk_device_protocol = fu_plugin_lookup_quirk_by_id (
+		plugin, quirk_key, DEVICE_PROTOCOL);
+	quirk_vendor_name = fu_plugin_lookup_quirk_by_id (
+                plugin, quirk_key, DEVICE_VENDOR_NAME);
+	/* Add devices with quirk configuration only. */
+	if (quirk_programmer_name == NULL)
+		return TRUE;
+
+	device_path = g_build_filename (
+		i2c_device_dir, i2c_name, NULL);
+	bus_no = get_i2c_bus_number_from_path (device_path, error);
+	if (bus_no == -1) {
```

Comment:
If this happens I think you end up overwriting the error that came from the function